nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/java/org/apache/nifi/json/TestJsonTreeRowRecordReader.java

```diff
@@ -726,4 +732,383 @@ public void testIncorrectSchema() throws IOException, MalformedRecordException {
+                        put("integer", 12);
+                        put("boolean", false);
+                        put("extraString", "extraStringValue12");
+                    }})
+                });
+            }}),
+            new MapRecord(expectedRecordChoiceSchema, new HashMap<String, Object>(){{
+                put("record", new Object[]{
+                    new MapRecord(expectedSelectedRecordSchemaForRecordsInBothArrays, new HashMap<String, Object>() {{
+                        put("integer", 21);
+                        put("extraString", "extraStringValue21");
+                        put("string", "stringValue21");
+                    }}),
+                    new MapRecord(expectedSelectedRecordSchemaForRecordsInBothArrays, new HashMap<String, Object>() {{
+                        put("integer", 22);
+                        put("extraString", "extraStringValue22");
+                        put("string", "stringValue22");
+                    }})
+                });
+            }})
+        );
+
+        // WHEN
+        // THEN
+        testReadRecords(jsonPath, schema, expected);
+    }
+
+    private void testReadRecords(String jsonPath, List<Object> expected) throws IOException, MalformedRecordException {
+        // GIVEN
+        final File jsonFile = new File(jsonPath);
+
+        try (
+            InputStream jsonStream = new ByteArrayInputStream(FileUtils.readFileToByteArray(jsonFile));
+        ) {
+            RecordSchema schema = inferSchema(jsonStream);
+
+            // WHEN
+            // THEN
+            testReadRecords(jsonStream, schema, expected);
+        }
+    }
+
+    private void testReadRecords(String jsonPath, RecordSchema schema, List<Object> expected) throws IOException, MalformedRecordException {
+        // GIVEN
+        final File jsonFile = new File(jsonPath);
+
+        try (
+            InputStream jsonStream = new ByteArrayInputStream(FileUtils.readFileToByteArray(jsonFile));
+        ) {
+            // WHEN
+            // THEN
+            testReadRecords(jsonStream, schema, expected);
+        }
+    }
+
+    private void testReadRecords(InputStream jsonStream, RecordSchema schema, List<Object> expected) throws IOException, MalformedRecordException {
+        // GIVEN
+        try (
+            JsonTreeRowRecordReader reader = new JsonTreeRowRecordReader(jsonStream, mock(ComponentLog.class), schema, dateFormat, timeFormat, timestampFormat);
+        ) {
+            // WHEN
+            List<Object> actual = new ArrayList<>();
+            Record record;
+            while ((record = reader.nextRecord()) != null) {
+                List<Object> dataCollection = Arrays.asList((Object[]) record.getValue("dataCollection"));
+                actual.addAll(dataCollection);
+            }
+
+            // THEN
+            List<Function<Object, Object>> propertyProviders = Arrays.asList(
```

Comment:
I find this logic very confusing, to have a List of functions that transform one object, chained together, and then to wrap an existing list with these transforms, just to compare them against the result of applying the same chain of transformations.

It would be far simpler to just transform the objects here and then make the assertion against the transformed objects.